File: PyPy3/백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
from collections import deque
import sys

a,b = map(int,sys.stdin.readline().split())
s = 0 #시작점
aa = [[] for i in range(a+1)]
for i in range(b):
    c,d = map(int,input().split())
    aa[c].append(d)
    aa[d].append(c)
    if i==b-1:
        s=d
visited = [False]*(a+1)

def bfs(aa,s):
    cnt = 0
    while visited.count(False)!=1:
        queue = deque([s])
        while queue:
            now = queue.popleft()
            if visited[now] == False:
                visited[now] = True
                for n in aa[now]:
                    if not visited[n]:
                        queue.append(n)
        cnt+=1
        for i in range(1,a+1):
            if visited[i]==False:
                s = i
                break
    return cnt

sys.stdout.write(str(bfs(aa,s)))

File: Java/백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.java
import java.util.*;

public class Main{
    public static int bfs(int[][] aa,int a){
        boolean[] visited = new boolean[a+1];
        int cnt = 0;
        int s = 0;
        while (true){
            for (int i=1;i<a+1;i++){
                if (visited[i]==false){
                    s=i;
                    break;
                }
                if(i==a){
                    return cnt;
                }
            }
            Queue<int[]> q = new LinkedList<>();
            q.add(new int[] {s});
            while (!(q.isEmpty())){
                int now = q.remove()[0];
                if (visited[now] == false){
                    visited[now] = true;
                    for (int i=1;i<a+1;i++){
                        if(aa[now][i]!=0 && visited[i]==false){
                            q.add(new int[] {i});
                        }
                    }
                }
            }
            cnt+=1;
        }
    }
    public static void main(String[] args){
        Scanner in = new Scanner(System.in);
        int a = in.nextInt();
        int b = in.nextInt();
        int s = 0;
        int[][] aa = new int[a+1][a+1];
        for (int i=0;i<b;i++){
            int c = in.nextInt();
            int d = in.nextInt();
            aa[c][d] = 1;
            aa[d][c] = 1;
        }
        System.out.println(bfs(aa,a));
    }
}